pad_server/src/state.c

```diff
@@ -1,44 +1,67 @@
-#include "state.h"
+#include <errno.h>
 #include <pthread.h>
+#include <semaphore.h>
+#include <stdio.h>
+
+#include "state.h"
 
 /* TODO: docs */
 void padstate_init(padstate_t *state) {
-    pthread_mutex_init(&state->lock, NULL);
+    pthread_rwlockattr_setkind_np(
+        &state->rw_lock_attr, PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP); // Linux only, change if we deploy to
```

Comment:
Change this to a standard option that is POSIX compliant. This way it will run on both Linux and NuttX.

Reply:
The problem is that on linux, without setting attr, the default is readers preference, while as far as I know on nuttx, attr doesn't actually do anything, and it's always writers preference. POSIX doesn't actually specify anything about readers preference or writers preference, just that it can be either. 

POSIX docs: https://pubs.opengroup.org/onlinepubs/7908799/xsh/pthread_rwlock_wrlock.html
Linux docs: https://man7.org/linux/man-pages/man3/pthread_rwlockattr_setkind_np.3.html 
NuttX docs: https://nuttx.apache.org/releases/7.21/ and https://github.com/apache/nuttx/blob/master/libs/libc/pthread/pthread_rwlock.c for attr not being used (not that much documentation for this that I could find sadly!). 

Let me know what you want me to! Also happy Thanksgiving!

Reply:
Okay so after looking at the documentation you provided, it seems that POSIX leaves most of the rwlock behaviour implementation dependent, including the attributes. So I would prefer you keep the attributes `NULL` (don't define them) since that's the most portable way to write this. If you want to, you could add an `#ifdef` macro to use the writer preference if compiling on Linux, but I honestly don't think it's worth the trouble because we'll get different behaviour on every platform regardless.